### PR TITLE
[Ability] Fully implement Sheer Force

### PR DESCRIPTION
**Pull request by frutescens**
## What are the changes the user will see?
If Meloetta happens to have the ability Sheer Force when using Relic Song, she does not transform.  
Serene Grace does not affect King's Rock proc chance anymore (Balance decision).

## Why am I making these changes?
Shareholder value up!

## What are the changes from a developer perspective?
Conditionals added to MeloettaFormChangePostMoveTrigger that check for Sheer Force. Tests also added. Serene Grace and Sheer Force tests modernized.

## How to test the changes?
1. Use Relic Song with a Meloetta that has the ability Sheer Force

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)

### CODE REVIEW

**Comment by Snailman11** [comment]
very small chance that it fixes
https://github.com/pagefaultgames/pokerogue/issues/2354
(Sheer force edge cases)

**Comment by innerthunder** [comment]
It'd be nice if you could rewrite the other Sheer Force tests along with this change. The tests right now tend to follow a bad pattern where they try to replicate their own code (applying the ability's attributes, etc.) and test against that output. If the components used aren't implemented correctly, the tests won't catch that, so the current tests hardly prove anything.

File: src/data/pokemon-forms.ts [hunk 1]
```diff
@@ -351,6 +351,10 @@ export class MeloettaFormChangePostMoveTrigger extends SpeciesFormChangePostMove
     if (pokemon.scene.gameMode.hasChallenge(Challenges.SINGLE_TYPE)) {
       return false;
     } else {
+      // Meloetta will not transform if it has the ability Sheer Force when using Relic Song
+      if (pokemon.hasAbility(Abilities.SHEER_FORCE, true)) {
```
Comment (innerthunder) [inline src/data/pokemon-forms.ts:355]:
```suggestion
      if (pokemon.hasAbility(Abilities.SHEER_FORCE)) {
```
`true` is the default value

File: src/modifier/modifier.ts [hunk 2]
```diff
@@ -732,11 +732,10 @@ export abstract class PokemonHeldItemModifier extends PersistentModifier {
     if (!pokemon.getLastXMoves()[0]) {
       return 1;
     }
-    const sheerForceAffected = allMoves[pokemon.getLastXMoves()[0].move].chance >= 0 && pokemon.hasAbility(Abilities.SHEER_FORCE);
-
+    const sheerForceAffected = pokemon.hasAbility(Abilities.SHEER_FORCE, true);
```
Comment (innerthunder) [inline src/modifier/modifier.ts:735]:
```suggestion
    const sheerForceAffected = pokemon.hasAbility(Abilities.SHEER_FORCE);
```

File: src/modifier/modifier.ts [hunk 3]
```diff
@@ -732,11 +732,10 @@ export abstract class PokemonHeldItemModifier extends PersistentModifier {
...
     if (sheerForceAffected) {
       return 0;
-    } else if (pokemon.hasAbility(Abilities.SERENE_GRACE)) {
+    } else if (pokemon.hasAbility(Abilities.SERENE_GRACE, true)) {
```
Comment (innerthunder) [inline src/modifier/modifier.ts:738]:
```suggestion
    } else if (pokemon.hasAbility(Abilities.SERENE_GRACE)) {
```

File: src/modifier/modifier.ts [hunk 4]
```diff
@@ -1403,7 +1403,10 @@ export class CritBoosterModifier extends PokemonHeldItemModifier {
    * @param critStage {@linkcode NumberHolder} that holds the resulting critical-hit level
    * @returns always `true`
    */
-  override apply(_pokemon: Pokemon, critStage: NumberHolder): boolean {
+  override apply(pokemon: Pokemon, critStage: NumberHolder): boolean {
+    if (super.getSecondaryChanceMultiplier(pokemon) === 0) {
+      return false;
+    }
```
Comment (innerthunder) [inline src/modifier/modifier.ts:1409]:
```suggestion
```
Scope Lens shouldn't be affected by Sheer Force

File: src/modifier/modifier.ts [hunk 5]
```diff
@@ -1449,7 +1453,7 @@ export class SpeciesCritBoosterModifier extends CritBoosterModifier {
    * @returns `true` if the critical-hit level can be incremented, false otherwise
    */
   override shouldApply(pokemon: Pokemon, critStage: NumberHolder): boolean {
-    return super.shouldApply(pokemon, critStage) && (this.species.includes(pokemon.getSpeciesForm(true).speciesId) || (pokemon.isFusion() && this.species.includes(pokemon.getFusionSpeciesForm(true).speciesId)));
+    return super.shouldApply(pokemon, critStage) && super.getSecondaryChanceMultiplier(pokemon) !== 0 && (this.species.includes(pokemon.getSpeciesForm(true).speciesId) || (pokemon.isFusion() && this.species.includes(pokemon.getFusionSpeciesForm(true).speciesId)));
```
Comment (innerthunder) [inline src/modifier/modifier.ts:1456]:
```suggestion
    return super.shouldApply(pokemon, critStage) && (this.species.includes(pokemon.getSpeciesForm(true).speciesId) || (pokemon.isFusion() && this.species.includes(pokemon.getFusionSpeciesForm(true).speciesId)));
```
Same deal as Scope Lens. These items are applied during attacks, not as a subsequent effect, so Sheer Force shouldn't negate their effects

File: src/modifier/modifier.ts [hunk 6]
```diff
@@ -2657,7 +2669,10 @@ export class PokemonMoveAccuracyBoosterModifier extends PokemonHeldItemModifier
    * @param moveAccuracy {@linkcode NumberHolder} holding the move accuracy boost
    * @returns always `true`
    */
-  override apply(_pokemon: Pokemon, moveAccuracy: NumberHolder): boolean {
+  override apply(pokemon: Pokemon, moveAccuracy: NumberHolder): boolean {
+    if (super.getSecondaryChanceMultiplier(pokemon)) {
+      return false;
+    }
```
Comment (innerthunder) [inline src/modifier/modifier.ts:2675]:
```suggestion
```
Same as the other items, not a post-attack effect

File: src/modifier/modifier.ts [hunk 7]
```diff
@@ -2692,6 +2710,9 @@ export class PokemonMultiHitModifier extends PokemonHeldItemModifier {
    * @returns always `true`
    */
   override apply(pokemon: Pokemon, moveId: Moves, count: NumberHolder | null = null, damageMultiplier: NumberHolder | null = null): boolean {
+    if (super.getSecondaryChanceMultiplier(pokemon) === 0) {
+      return false;
+    }
```
Comment (innerthunder) [inline src/modifier/modifier.ts:2715]:
```suggestion
```

**Review by innerthunder — CHANGES_REQUESTED** [review]
There are a few items incorrectly being cancelled out here...

**Review by innerthunder — DISMISSED** [review]

File: src/modifier/modifier.ts [hunk 8]
```diff
@@ -1644,7 +1630,8 @@ export class FlinchChanceModifier extends PokemonHeldItemModifier {
    * @returns `true` if {@linkcode FlinchChanceModifier} has been applied
    */
   override apply(pokemon: Pokemon, flinched: BooleanHolder): boolean {
-    if (!flinched.value && pokemon.randSeedInt(10) < (this.getStackCount() * this.getSecondaryChanceMultiplier(pokemon))) {
+    // The check for pokemon.battleSummonData is to ensure that a crash doesn't occur when a Pokemon with King's Rock procs a flinch
```
Comment (DayKev) [inline src/modifier/modifier.ts:1633]: There's no check for `pokemon.battleSummonData` though?

File: src/data/ability.ts [hunk 9]
```diff
@@ -5709,9 +5709,7 @@ export function initAbilities() {
       .condition(getSheerForceHitDisableAbCondition()),
     new Ability(Abilities.SHEER_FORCE, 5)
       .attr(MovePowerBoostAbAttr, (user, target, move) => move.chance >= 1, 5461 / 4096)
-      .attr(MoveEffectChanceMultiplierAbAttr, 0)
-      .edgeCase() // Should disable shell bell and Meloetta's relic song transformation
```
Comment (PigeonBar) [inline src/data/ability.ts:5713]: Just realized that we forgot to also add the shell bell interaction to this PR

File: src/modifier/modifier.ts [hunk 10]
```diff
@@ -1644,15 +1634,16 @@ export class FlinchChanceModifier extends PokemonHeldItemModifier {
...
+    if (pokemon.battleSummonData && !flinched.value && pokemon.randSeedInt(100) < (this.getStackCount() * this.chance)) {
       flinched.value = true;
       return true;
```
Comment (PigeonBar) [inline src/modifier/modifier.ts:1640]:
```suggestion
      const sheerForceAffected = allMoves[pokemon.getLastXMoves()[0].move].chance >= 0 && pokemon.hasAbility(Abilities.SHEER_FORCE);
      if (!sheerForceAffected) {
        flinched.value = true;
        return true;
      }
```
Assuming that Sheer Force activation should still disable King's Rock

Comment (frutescens) [reply src/modifier/modifier.ts:1640]: Ah, no. It will be separate from King's Rock.

File: src/data/ability.ts (hunk 9 above)
Comment (frutescens) [reply src/data/ability.ts:5713]: Ah, this was my bad for not updating the description but Balance decided to have item effects separate from moves so shell bell will not be affected by sheer force + shell bell will not boost the BP of moves when held by the pokemon. Same with King's Rock, etc.

**Review by innerthunder — APPROVED** [review]

**Comment by frutescens** [comment]
image

**Pull request merged**